packages/next-intl/src/middleware/utils.tsx

```diff
@@ -9,6 +9,71 @@ export function getFirstPathnameSegment(pathname: string) {
   return pathname.split('/')[1];
 }
 
+function getExternalPath<Locales extends AllLocales>(
+  path: string | {[Key in Locales[number]]: string}
+): string {
+  const firstLocalePath = Object.values(path)[0];
+  return typeof path === 'string' ? path : firstLocalePath;
+}
+
+function isCatchAllRoute(pathname: string) {
+  return isOptionalCatchAll(pathname) || isCatchAll(pathname);
+}
+
+function isOptionalCatchAll(pathname: string) {
+  return pathname.includes('[[...');
+}
+
+function isCatchAll(pathname: string) {
+  return pathname.includes('[...');
+}
+
+function isDynamicRoute(pathname: string) {
+  return pathname.includes('[');
+}
+
+type PathnamePair<Locales extends AllLocales> = [
+  string,
+  string | {[Key in Locales[number]]: string}
+];
+
+export function comparePathnamePairs<Locales extends AllLocales>(
+  a: PathnamePair<Locales>,
+  b: PathnamePair<Locales>
+): number {
+  const pathA = getExternalPath(a[1]);
+  const pathB = getExternalPath(b[1]);
+
+  const dynamicA = isDynamicRoute(pathA);
+  const dynamicB = isDynamicRoute(pathB);
+  const catchAllA = isCatchAllRoute(pathA);
+  const catchAllB = isCatchAllRoute(pathB);
```

Comment:
Diving a bit deeper into this, I'm not sure these flags are sufficient to be honest.

E.g. consider this case:
```
/articles/[category]/[articleSlug]
/articles/[category]/just-in
```

In this case, the routes would be considered equal in regard to priority, and probably the natural ordering would be preserved. However, the second route is more specific and should match.

The number of dynamic segments could be another indicator, but I think this won't solve the problem cleanly as e.g. `/foo/[d1]` and `/foo/bar/[d1]` have the same number of segments.

It seems like to be able to tell which route is really more specific, we have to consider routes in a hierarchical way, e.g. find the most specific route in `/articles`, then try to find the most specific route in `./[category]`, etc.

Do you by chance know if Next.js handles these cases correctly? Maybe we can take some inspiration from them.

What do you think?

Reply:
Or maybe the number of dynamic segments could in fact suffice? 🤔 

In the example `/foo/[d1]` and `/foo/bar/[d1]`, the routes have the same priority, but only one of them matches eventually—so maybe this isn't an issue.

Do you think you could a few more tests for edge cases so we can gain some clarity about those?

Reply:
Ah yes, thanks for the hint! You were absolutely right, my solution didn't take into account the case where we have nested dynamic routes, e.g: `/articles/[category]/[slug]` and a static route in the same place 😅.

I'm not very familiar with the Next.js repo, but at first glance I believe the code you referenced in the corresponding issue is actually where Next does the route sorting:
https://github.com/vercel/next.js/blob/415cd74b9a220b6f50da64da68c13043e9b02995/packages/next/src/shared/lib/router/utils/sorted-routes.ts

When looking at Next's code I realized that it is necessary that the individual route segments are compared with each other and have now implemented this additionally so that the failed tests that I added at the beginning of this update now pass.

If there are any other edge cases you can think of, I'm happy to add more tests! 🙂